service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/AzureVmHelper.java

```diff
@@ -44,6 +49,73 @@ public static StepResult deleteVm(
     return StepResult.getStepResultSuccess();
   }
 
+  public static StepResult removeAllUserAssignedManagedIdentitiesFromVm(
+      AzureCloudContext azureCloudContext, ComputeManager computeManager, String vmName)
+      throws InterruptedException {
+    try {
+      Set<String> userAssignedMsiIds =
+          computeManager
+              .virtualMachines()
+              .getByResourceGroup(azureCloudContext.getAzureResourceGroupId(), vmName)
+              .userAssignedManagedServiceIdentityIds();
+
+      if (userAssignedMsiIds != null) {
+        userAssignedMsiIds.forEach(
+            (String userAssignedMsiId) -> {
+              computeManager
```

Comment:
Wonder if we can avoid fetching the VM multiple times:
```
VirtualMachine resolvedVm = computeManager
    .virtualMachines()
    .getByResourceGroup(azureCloudContext.getAzureResourceGroupId(), vmName)
```

Then call `resolvedVm. userAssignedManagedServiceIdentityIds()`, `resolvedVm.update(..)`.